gobblin-modules/google-ingestion/src/main/java/gobblin/ingestion/google/adwords/GoogleAdWordsReportDownloader.java

```diff
@@ -0,0 +1,379 @@
+
+    _dailyPartition = state.getPropAsBoolean(GoogleAdWordsSource.KEY_CUSTOM_DATE_DAILY, false);
+    _debugFileOutputPath = state.getProp(GoogleAdWordsSource.KEY_DEBUG_PATH_FILE, "");
+    _debugStringOutputPath = state.getProp(GoogleAdWordsSource.KEY_DEBUG_PATH_STRING, "");
+    _maxThreadsAllowed = state.getPropAsInt(GoogleAdWordsSource.KEY_THREADS, 8);
+
+    _skipReportHeader = state.getPropAsBoolean(GoogleAdWordsSource.KEY_REPORTING_SKIP_REPORT_HEADER, true);
+    _skipColumnHeader = state.getPropAsBoolean(GoogleAdWordsSource.KEY_REPORTING_SKIP_COLUMN_HEADER, true);
+    _skipReportSummary = state.getPropAsBoolean(GoogleAdWordsSource.KEY_REPORTING_SKIP_REPORT_SUMMARY, true);
+    _useRawEnumValues = state.getPropAsBoolean(GoogleAdWordsSource.KEY_REPORTING_USE_RAW_ENUM_VALUES, false);
+    _includeZeroImpressions = state.getPropAsBoolean(GoogleAdWordsSource.KEY_REPORTING_INCLUDE_ZERO_IMPRESSION, false);
+  }
+
+  static List<String> schemaToColumnNames(JsonArray schemaJsonArray) {
+    List<String> fields = new ArrayList<>();
+    for (int i = 0; i < schemaJsonArray.size(); i++) {
+      fields.add(schemaJsonArray.get(i).getAsJsonObject().get("columnName").getAsString());
+    }
+    return fields;
+  }
+
+  public void downloadAllReports(Collection<String> accounts, final LinkedBlockingDeque<String[]> reportRows)
+      throws InterruptedException {
+    ExecutorService threadPool = Executors.newFixedThreadPool(Math.min(_maxThreadsAllowed, accounts.size()));
+    List<Pair<String, String>> dates = getDates();
+    Map<String, Future<Void>> jobs = new HashMap<>();
+
+    for (String acc : accounts) {
+      final String account = acc;
+      for (Pair<String, String> dateRange : dates) {
+        final Pair<String, String> range = dateRange;
+        final String jobName;
+        if (_dateRangeType.equals(ReportDefinitionDateRangeType.ALL_TIME)) {
+          jobName = String.format("'all-time report for %s'", account);
+        } else {
+          jobName = String.format("'report for %s from %s to %s'", account, range.getLeft(), range.getRight());
+        }
+
+        Future<Void> job = threadPool.submit(new Callable<Void>() {
+          @Override
+          public Void call()
+              throws Exception {
+            Callable<Void> downloadJob = new Callable<Void>() {
+              @Override
+              public Void call()
+                  throws ReportDownloadResponseException, InterruptedException, IOException, ReportException {
+                log.info("Start downloading " + jobName);
+                downloadReport(account, range.getLeft(), range.getRight(), reportRows);
+                log.info("Successfully downloaded " + jobName);
+                return null;
+              }
+            };
+            _retryer.call(downloadJob);
+            return null;
+          }
+        });
+
+        jobs.put(jobName, job);
+        Thread.sleep(100);
```

Comment:
This is flaky. The download might not have completed by the time primary thread moves on. You should instead be blocking on future.get()